### ISSUE TITLE
Trusted Type error when launching extension with 'browser' entry point in node host

### ISSUE THREAD

**Issue by Elringus** [microsoft/vscode#138413]
Issue Type: <b>Bug</b>

When attempting to launch a web-compatible extension with `browser` entry point under node VS Code host, following error is thrown:

![](https://i.gyazo.com/4e778eed5ef865090b330135c626063d.png)

When `main` entry point is specified (in addition to `browser`), the error is not thrown, but I need the extension to run in browser environment to access `WebSocket` API (it uses .NET Blazor, which doesn't work with node's `ws` module).

Here is the sample extension sources: https://github.com/Elringus/DotNetJS/tree/main/Samples/WebExtension

VS Code version: Code 1.62.3 (ccbaa2d27e38e5afa3e5c21c1c7bef4657064247, 2021-11-17T08:11:14.551Z)
OS version: Windows_NT x64 10.0.19042
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6900K CPU @ 3.20GHz (16 x 3198)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|63.92GB (48.88GB free)|
|Process Argv|-n --crash-reporter-id 5553deaa-2949-4706-8450-f98d274a63e4|
|Screen Reader|no|
|VM|50%|
</details><details><summary>Extensions (4)</summary>

Extension|Author (truncated)|Version
---|---|---
dotnetjs|Elr|1.0.6
naninovel|Elr|1.6.1
vscode-dotnet-runtime|ms-|1.5.0
code-spell-checker|str|2.0.13

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythontb:30283811
pythonvspyt551cf:30345471
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
pythonrunftest32:30373476
pythonf5test824:30373475
javagetstartedt:30391933
pythonvspyt187:30373474
vsaa593cf:30376535
pythonvs932:30405811
vscexrecpromptt1:30404947
vscop804:30404766
vscop341:30404997
vsrem710:30405998

```

</details>

**Comment by Elringus** [comment]
And here is the published extension in question, in case you'd like to test it: https://marketplace.visualstudio.com/items?itemName=Elringus.dotnetjs It works fine in browser (via vscode.dev), but throws the error in standalone VS Code.

**Comment by jrieken** [comment]
VS Code uses [trusted types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/trusted-types) as additional measure to protect from XSS. This is defined per document(-header). When running in web, e.g gh.dev or vscode.dev, we isolate the web-worker running extensions in a separate iframe and that iframe doesn't inherit nor use trusted types. In contrast, when running in desktop we don't further isolate the web worker and therefore  trusted types restrictions apply to extensions too

**Comment by Elringus** [comment]
Thanks for the response! I've checked the linked article, but still not sure how to solve the issue. Should I add a trusted type policy before the extension is activated? If that's the case, what exactly should be inside the policy object for it to work in VS Code? It complains about evaluating a string as JavaScript code, so I guess it's about `eval()`, but the extension doesn't use eval, nor does it use any of the DOM APIs, which are governed by the trusted types.

**Comment by jrieken** [comment]
Yeah, I guess it's about `eval`, likely inside library or generated code. We don't allow that in our code and in the web_worker+desktop case that's a restriction that you are inherited. I believe that's unwanted strictness from our side and something we should relax.

**Comment by Elringus** [comment]
There are indeed 3 eval calls in the generated code, but they're not executed. Just to make sure, I've replaced all of them with `{}`, but the same error is still thrown in VS Code.

**Comment by Elringus** [comment]
@jrieken @alexdima Thanks for fixing this so quickly! Can you please clarify if there are still some limitations on using `WebSocket` API in web extensions running in standalone? I can't connect to local websocket server (`ws://localhost`) in standalone mode, while in browser it works fine.

**Comment by jrieken** [comment]
Even locally we are strict and want `wss` (note the secure `s`)

**Comment by Elringus** [comment]
> Even locally we are strict and want `wss` (note the secure `s`)
Ah, I see. Thought about it, but the fact that unsecure connection is working in browser and in standalone with .NET runtime gave me hope that you could allow it in this case as well. Thanks for the quick response!
